### PR TITLE
Added kafka streaming source offset and partition options

### PR DESCRIPTION
**Pull request by Luishfs**
# Description

This PR contemplates three issues (#4069 #4114 #4040)

## Offset handling
Kafka source will now allow for offset configuration along with 
partition settings.
These offsets being:

1. Beginning
2. End
3. Int
4. Timestamp

This allows users to set specific positions inside a topic to consume data

offset handling corresponds to #4114 and #4069

## Auto Offset reset
in #4069 user requested a similar behaviour to the auto_offset_reset python function
The specified function in the issue seems to be from Spark.
With that, the whole content of the issue is solved inside Offset Handling, but this 
setting was added to prevent future issues

## Max Partition fetch bytes
Added max_partition_fetch_bytes settings along with auto_offset_reset.

this corresponds to #4040

# How Has This Been Tested?

Tested using a local kafka broker and Dummy sink.

## Beginning offset
For this test, 10 messages were written with a producer, counting from 1 to 10 **before** running the pipeline
**Expected behaviour**: All 10 messages are displayed 

Screenshot from 2023-12-19 18-19-54

Screenshot from 2023-12-19 18-22-27

**Result**: All 10 messages were displayed

## End offset
For this test, 10 messages were written with a producer, counting from 1 to 10 **before** running the pipeline,
And one message (11) **after** running the pipeline.
**Expected behaviour**: Only the message containing 11 is displayed

Screenshot from 2023-12-19 18-39-20

Screenshot from 2023-12-19 18-40-02

**Result**: The message containing 11 is displayed

## Int offset
For this test, 12 messages were written with a producer, counting from 1 to 12 **before** running the pipeline
Only one partition is being used to simplify the expected behaviour
**Expected behaviour**: Only the messages from and after 1:12 (partition number and position) should be displayed

Screenshot from 2023-12-19 18-52-15

Screenshot from 2023-12-19 18-52-24

**Result**: The message from 1:12 is displayed

## Timestamp offset
For this test, 11 messages were written with a producer, counting from 1 to 11 **before** running the pipeline
Only one partition is being used to simplify the expected behaviour
**Expected behaviour**: Only the messages from and after the given timestamp should be displayed
( the selected timestamp corresponds to the last message timestamp ) 

Screenshot from 2023-12-19 19-08-58

Screenshot from 2023-12-19 19-09-06

**Result**: The message from the given timestamp is displayed

# Checklist
- [X] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation

cc:

@wangxiaoyou1993

### CODE REVIEW

**Comment by wangxiaoyou1993** [comment]
cc: @sujiplr

File: docs/streaming/sources/kafka.mdx [hunk 1]
```diff
@@ -14,8 +14,30 @@ include_metadata: false
 api_version: "0.10.2"
 batch_size: 100
 timeout_ms: 500
+auto_offset_reset: latest # Optional
```
Comment (wangxiaoyou1993) [inline docs/streaming/sources/kafka.mdx:17]: what are the possible values for auto_offset_reset?

File: docs/streaming/sources/kafka.mdx [hunk 2]
```diff
@@ -14,8 +14,30 @@ include_metadata: false
...
+max_partition_fetch_bytes: 1048576 # Optional
+offset: None # Optional
```
Comment (wangxiaoyou1993) [inline docs/streaming/sources/kafka.mdx:19]: You can put these fields all under offset as a dictionary
```
offset:
  type:
  value:
  partitions: 
```

File: mage_ai/streaming/sources/kafka.py [hunk 3]
```diff
@@ -42,13 +42,18 @@ class KafkaConfig(BaseConfig):
     api_version: str = '0.10.2'
     batch_size: int = DEFAULT_BATCH_SIZE
     timeout_ms: int = DEFAULT_TIMEOUT_MS
+    auto_offset_reset: str = 'latest'
```
Comment (wangxiaoyou1993) [inline mage_ai/streaming/sources/kafka.py:45]: where do you use this field?

File: docs/streaming/sources/kafka.mdx (hunk 1 above)
Comment (Luishfs) [reply docs/streaming/sources/kafka.mdx:17]: There are only 2 possible values: latest & earliest
If the consumer faces a issue reguarding offsets ( e.g OffsetOutOfRange), the consumer will reset its offset based on the given value.
earliest meaning the oldest message possible
latest meaning the most recent message possible

File: mage_ai/streaming/sources/kafka.py (hunk 3 above)
Comment (Luishfs) [reply mage_ai/streaming/sources/kafka.py:45]: This field is used in [consumer_kwargs](https://github.com/mage-ai/mage-ai/pull/4226/files#diff-d183d9cbac5329d37f58e32ec12a23789222c7be7978f22dca04286bb353e6daR88)  to initialize a kafka consumer

File: docs/streaming/sources/kafka.mdx (hunk 2 above)
Comment (Luishfs) [reply docs/streaming/sources/kafka.mdx]: Modified in 9b890bb8c9df6a8549ce684c5edf72273c42e2e2

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**